crates/hotshot/src/tasks/mod.rs

```diff
@@ -364,6 +366,83 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> EventTransformerState<TYPES,
     }
 }
 
+#[derive(Debug)]
+/// An `EventHandlerState` that modifies justify_qc to that of previous view to mock dishonest leader
+pub struct DishonestLeader<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+    /// store events from previous views
+    cached_events: Vec<HotShotEvent<TYPES>>,
+    /// phantom
+    _phantom: std::marker::PhantomData<I>,
+}
+
+/// add trait to handle proposal events for a dishonest leader
+trait DishonestProposal<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+    /// send dishonest proposal
+    fn handle_proposal_send_event(
+        &self,
+        event: &HotShotEvent<TYPES>,
+        proposal: &Proposal<TYPES, QuorumProposal<TYPES>>,
+        sender: &TYPES::SignatureKey,
+    ) -> Vec<HotShotEvent<TYPES>>;
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> Default for DishonestLeader<TYPES, I> {
+    /// create a dishonest leader that stores past events to try a modify a view when they are the leader
+    fn default() -> Self {
+        DishonestLeader {
+            cached_events: Vec::new(),
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> DishonestProposal<TYPES, I>
+    for DishonestLeader<TYPES, I>
+{
+    fn handle_proposal_send_event(
+        &self,
+        event: &HotShotEvent<TYPES>,
+        proposal: &Proposal<TYPES, QuorumProposal<TYPES>>,
+        sender: &TYPES::SignatureKey,
+    ) -> Vec<HotShotEvent<TYPES>> {
+        let dishonest = self.cached_events.len() == 2;
+        if !dishonest {
+            return vec![event.clone()];
+        }
```

Comment:
IIUC this means this leader will send 1 dishonest proposal on the second time it proposes.

Also you don't need a variable for `dishonest`

Reply:
yes, that is what it currently does. the second time this node node proposes it will send old qc